src/protocol/attribution/mod.rs

```diff
@@ -5,82 +5,96 @@ pub mod credit_capping;
+    validator: V,
+    binary_validator: VB,
     arithmetically_shared_values: Vec<ArithmeticallySharedIPAInputs<F, S>>,
     binary_shared_values: Vec<BinarySharedIPAInputs<SB>>,
     config: IpaQueryConfig,
-) -> Result<Vec<MCAggregateCreditOutputRow<F, SemiHonestAdditiveShare<F>, BK>>, Error>
+) -> Result<Vec<Replicated<F>>, Error>
 where
-    C: UpgradableContext,
-    C::UpgradedContext<F>: UpgradedContext<F, Share = S> + RandomBits<F, Share = S>,
-    S: LinearSecretSharing<F> + BasicProtocols<C::UpgradedContext<F>, F> + Serializable + 'static,
+    V: Validator<C, F>,
+    VB: Validator<C, Gf2>,
+    C: UpgradableContext<Validator<F> = V>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F>
+        + BasicProtocols<C::UpgradedContext<F>, F>
+        + Serializable
+        + DowngradeMalicious<Target = Replicated<F>>
+        + 'static,
     C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = SB> + Context,
     SB: LinearSecretSharing<Gf2>
         + BasicProtocols<C::UpgradedContext<Gf2>, Gf2>
         + DowngradeMalicious<Target = Replicated<Gf2>>
         + 'static,
     F: PrimeField + ExtendableField,
-    BK: GaloisField,
     ShuffledPermutationWrapper<S, C::UpgradedContext<F>>: DowngradeMalicious<Target = Vec<u32>>,
-    MCAggregateCreditOutputRow<F, S, BK>:
-        DowngradeMalicious<Target = MCAggregateCreditOutputRow<F, SemiHonestAdditiveShare<F>, BK>>,
 {
+    let row_count = arithmetically_shared_values.len();
+    assert_eq!(row_count, binary_shared_values.len());
     let m_ctx = validator.context();
     let m_binary_ctx = binary_validator.context();
 
+    // There are one fewer helper bits than there are input rows.  Same for stop bits.
+    // This propagates throughout aggregation (all the code understands this).
+    // And a breakdown for the last row isn't necessary because an impression on that row can't convert.
+    // So we drop the last breakdown key right away.
     let helper_bits_gf2 = compute_helper_bits_gf2(m_binary_ctx, &binary_shared_values).await?;
     let breakdown_key_bits_gf2: Vec<_> = binary_shared_values
         .iter()
         .map(|x| x.breakdown_key.clone())
+        .take(row_count - 1)
         .collect();
+
     let (validated_helper_bits_gf2, validated_breakdown_key_bits_gf2) = binary_validator
         .validate((helper_bits_gf2, breakdown_key_bits_gf2))
         .await?;
-    let semi_honest_fp_helper_bits =
-        mod_conv_helper_bits(ctx.clone(), &validated_helper_bits_gf2).await?;
-    let helper_bits = once(S::ZERO)
-        .chain(m_ctx.upgrade(semi_honest_fp_helper_bits).await?)
-        .collect::<Vec<_>>();
+
+    let convert_ctx = m_ctx
+        .narrow(&AttributionStep::ConvertHelperBits)
+        .set_total_records(validated_helper_bits_gf2.len());
+    let helper_bits = convert_bits(convert_ctx, stream_iter(validated_helper_bits_gf2), 0..1)
+        .map_ok(|b| b.into_iter().next().unwrap())
```

Comment:
The type we have here is `b: BitDecomposed`.  This contains a single value (as we're converting a single bit).  And because it is a wrapper, the number of options we have for unpacking it are few.  I could have added one, but this worked and I didn't think much more on it.

This can probably be improved - we have a few places where a single bit is upconverted like this and it is probably worth putting this into a wrapper function.  I'll flag it with a TODO.